Modal component

Ghrehh opened the pull request:
## Description
Adds a Modal component to the client with functionality similar to the existing sidepanel.

## Addresses
- https://linear.app/budibase/issue/BUDI-8242/modal-component

## Screenshots
|description|picture|
|---|---|
|Small Min|![Screenshot 2024-06-04 at 10 55 52](https://github.com/Budibase/budibase/assets/12501626/b121ba2a-d1d4-4e13-beb9-83d902309e58)|
|Small Scroll|![Screenshot 2024-06-04 at 10 56 01](https://github.com/Budibase/budibase/assets/12501626/6dfb18ed-e217-45cb-872f-4c2bd7d46dae)|
|Med Min|![Screenshot 2024-06-04 at 10 56 22](https://github.com/Budibase/budibase/assets/12501626/06a5a4a1-9136-4ca0-81e0-1036c308867f)|
|Med Scroll|![Screenshot 2024-06-04 at 10 56 27](https://github.com/Budibase/budibase/assets/12501626/1dd8f9a9-e4a8-44ca-9bf7-ba0feb6cd2bc)|
|Large Min|![Screenshot 2024-06-04 at 10 56 41](https://github.com/Budibase/budibase/assets/12501626/6c630770-52a9-4cc8-8cdb-7589756ceeb6)|
|Large Scroll|![Screenshot 2024-06-04 at 10 56 48](https://github.com/Budibase/budibase/assets/12501626/3863a3ed-e095-449d-a064-03d9489f1e66)|
|Fullscreen Min|![Screenshot 2024-06-04 at 10 56 54](https://github.com/Budibase/budibase/assets/12501626/1dce66a9-4f22-43fc-bf48-207d431686a1)|
|Fullscreen Scroll|![Screenshot 2024-06-04 at 10 57 00](https://github.com/Budibase/budibase/assets/12501626/87414aac-e181-4e0f-b2c7-7d900425bddc)|

joebudi commented:
@Ghrehh could you share an image of the settings panel for the modal?

deanhannigan reviewed (changes requested):
🎉  Fantastic work Gerard! A long overdue feature and it looks awesome! The code itself looks grand and I've outlined some issues I ran into during testing below:

`BUG` - Clicking an item in the tree from within the modal seems to cause a reactivity loop that kills the tab. This seemed to manifest when I added a formblock
`BUG` - User prompt actions in the `onClose` behaviour are terminated when the modal closes. The action type should possibly be filtered out for modals or, ideally, suspend the close.

`BUG` - Seems like theming isn't applied to the modal or its contents. 
![Screenshot 2024-06-05 at 10 25 21](https://github.com/Budibase/budibase/assets/5913006/f62dbb72-aa08-44a8-89aa-82abd8bd2fb0)
`BUG` - Error in the console when clicking outside the modal or attempting to click close, in builder preview. Outlined below.
`BUG` - Scroll bar in the modal overlay when not scrollable. Outlined below.
`BUG` - Contents not resizing in the builder preview. I added a chart to a small modal and then updated it to large. The chart did not resize in the builder preview but did display as intended on refresh
`NOTE` - Missing dotted outline when empty. This is specified in the ticket.
`NOTE` - Deleting anything in the modal cause it to flicker.
`NOTE` - Deleting the last component in a modal will cause the modal to close. The current tree behaviour selects the next node regardless of where it is. In this instance it should possibly stick to the last component in the modal or the modal itself?

aptkingston commented:
Just a heads up that I'm waiting for that initial feedback to be addressed before reviewing, just to save time discoering and reporting the same issues. Just in case you were waiting for my review before making changes.

Ghrehh commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,147 @@
+
+  const handleModalClose = async () => {
+    modalStore.actions.close()
+    await onClose()
```
Was a little confused by this at first, as I thought my default argument for the prop of an empty async function would be sufficient. But turns out if you add an on-close action, then remove it, you get the error you mentioned.

Maybe my svelte-fu is lacking, because I'd assume the parent of this component is passing something nullish to this prop, which should then use the default 🤷 Maybe it's a bound prop or something, idk, but good catch, I'll add the guard.

aptkingston replied:
Svelte default props work the same as default function params - they only get used if the value is `undefined`, but not for `null`. Just to explain why that's probably happening.

Ghrehh commented:
@deanhannigan Cheers for the thorough review. Addressed your feedback with the exception of:

`BUG - Contents not resizing in the builder preview. I added a chart to a small modal and then updated it to large. The chart did not resize in the builder preview but did display as intended on refresh`

This seems to happen for other, existing fixed-sized components too, like if you have a fixed width `container` with a chart inside it, changing the width of that container won't update the size of chart. This might be a chart only issue, as components like `text` resize appropriately in both `container`s and `modal`s. So this is definitely something worth addressing, but since I didn't introduce it would you be okay with us just ticketing it for later?


`NOTE - Deleting anything in the modal cause it to flicker.`

If I'm seeing what I think you're seeing, I think this might be caused by the component settings menu rapidly opening and closing as the next appropriate component is selected, this does look janky, but also happens currently in master for any component, so again might be worth punting, let me know, I might be misunderstanding what you mean.

deanhannigan reviewed (changes requested):
Most of the original issues have been resolved! I've raised some minor issues below, but other than that, the modal component is looking great!

aptkingston commented on packages/client/src/components/ClientApp.svelte:
```diff
@@ -128,6 +128,7 @@
     class:builder={$builderStore.inBuilder}
     class:show={fontsLoaded && dataLoaded}
   >
+    <div class="modal-portal" />
```
We already have a `modal-container` node further down (line 210) that we use for all normal modals. Can we not just use that here too?
We would just need to update the class name inside your new modal component as it also uses the same class name.

aptkingston commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,162 @@
+
+  const handleModalClose = async () => {
+    if (onClose) {
+      await onClose()
```
That sort of feature will tie in to something I've already been working on, which is prompting users before navigating away to avoid losing work, but I don't think we should make any changes for that at the moment.

aptkingston commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,162 @@
+    >
+      <div class="modal-container">
+        <div
+          transition:fly={{ duration: 300, y: 300 }}
```
Agreed, we should use the same values as the BBUI modal here (duration 200, y 30).

aptkingston commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,162 @@
+{#if open}
+  <Portal target=".modal-portal">
+    <div
+      transition:fade={{ duration: 300 }}
```
We should use 200 durations across the board to stay consistent with the BBUI modal.

aptkingston commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,162 @@
+
+<style>
+  .modal-scroll-container {
+    background-color: #00000078;
```
Agreed, we should use the same styles as our BBUI modal as much as possible.

Ghrehh replied:
I was just eyeballing it based on the mockups in the ticket, if anything I'd say maybe the current overlay is a little lighter compared to the mockup.

aptkingston replied:
Consistency is the most important thing, so lets just use everything we currently use in the existing modal (otherwise users will see 2 different styles of modals depending on whether they're using this component or seeing a default modal for a confirmation prompt etc).

Ghrehh commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  .modal {
+    background-color: var(--spectrum-global-color-gray-50);
```
I based it off the current style of the side panel, but looking again at the mockups I think the correct color is `--spectrum-alias-background-color-primary` and i will change it to that.

aptkingston reviewed (commented):
Nice work @Ghrehh! Works really well.

Most of those comments of mine are saying the same thing - I think we should just copy as much of the core BBUI modal styles as possible. All that CSS (and required div layering) was already worked out and written there, so I think we should aim to just copy a lot of it.

Ghrehh commented:
@aptkingston @deanhannigan I'll change the styles to be the same as the BBUI modal, I was going off the mockup before.

aptkingston commented:
> @aptkingston @deanhannigan I'll change the styles to be the same as the BBUI modal, I was going off the mockup before.

Understandable of course, but I think consistency is more important. So thanks for doing that.

Ghrehh commented:
@aptkingston @deanhannigan If you guys could give this another look when you get a chance please, I've addressed your feedback and made the modal look like the one used in the builder

Ghrehh commented:
@deanhannigan Fixed the modal background colour that you pointed out in standup, good catch, ty 🙏

aptkingston commented:
> @deanhannigan Fixed the modal background colour that you pointed out in standup, good catch, ty 🙏

You'll just need to add `opacity: 0.65` as well - right now it's a solid background colour. That won't be as easy here because the scroll container actually contains components. In the BBUI modal we deliberately have a dedicated background layer which we apply this opacity to. I think you'll need to follow the BBUI approach here and add a dedicated underlay.

I mentioned it before, but I still think just copying the outer few skeleton divs of the BBUI modal might be the best approach because we had already identified and solved most of these problems there. That includes things like just using `on:mousedown|self` rather than `clickOutside`, which works well in this case since the underlay covers the full screen. But obviously feel free to just rejig this a bit to add a dedicated underlay layer rather than more significant changes at this stage.

Ghrehh commented:
@aptkingston 
Just noticed the opacity thing myself when trying to write a ticket for a different issue, that literally worked until I swapped branches 🙃, I tried every theme and all, absolutely no idea why, I promise I'm not a moron.

And I'll implement the other change you suggested also

Ghrehh commented:
@aptkingston Feedback addressed if you could look at this again please, I just used the bbui modal component, but I needed to add a new prop to it to override the margin, as the default one was way different from what the mockups in the ticket were suggesting. But Idk if we even care about that any more, so I can remove it if you want.

deanhannigan reviewed (approved):
LGTM! Nicely done Gerard! Retested there and the Modal's in great shape

aptkingston commented on packages/client/src/components/app/Modal.svelte:
```diff
@@ -0,0 +1,139 @@
+  on:cancel={handleModalClose}
+  bind:this={modal}
+  disableCancel={$builderStore.inBuilder}
+  zIndex={2}
```
Purely out of curiosity - what's the z index of 2 for?

Ghrehh replied:
This puts it above the app header which seems to be `z-index: 1` but below the selection indicators.

the default `999` value the modal has for this makes it so that the selection/highlight indicators for the modal appear under the translucent overlay.

aptkingston reviewed (approved):
Looks great! I think using the BBUI modal is a great call - it ensures consistency and reduces duplication and maintenance overhead.

In terms of the margin, personally I prefer the larger 40px margin so I would probably just change the BBUI modal to be 40 on both sides and let things be consistent again because I don't think the client and builder need different modal margins. But it's not exactly a blocker.

One other change which would be good to add is closing the modal whenever navigation happens. This is a new feature for side panels that Conor just shipped. Just needs one line added to ClientApp.svelte at line 110 to just also close the modal after it closes the side panel in `handleHashChange`.

Ghrehh commented:
@aptkingston Addressed your feedback

- Changed the 40px top and bottom margin to be 40px padding on the next container up (40px margin on all sides was overflowing weirdly due to the nested flex-boxes) which does the same job. Removed the customMargin prop.

- Added the call to close the modal on navigation like you requested

There was also an issue with elements in a modal that were deleted still being visible in the preview. This only appeared after swapping to use the bbui modal, and seems to be because the `Portal` is always rendered, but the children of it are conditional. Adding a guard to display/remove the portal based on visibility (in the builder only) fixed the issue. 
 
Behaviour for hiding/showing elements in the client proper seems fine without the above mentioned guard, hence why it's only done in the builder. I'm not sure what is the underlying cause of this is, but it seems to be fixed.
